In [6]:
%load_ext autoreload
# Import necessary libraries
import pandas as pd
import numpy as np
import sys
import os
sys.path.append(os.path.abspath('..'))
from scripts.ab_hypothesis_testing import *

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [13]:
%autoreload 2

In [8]:
df=pd.read_csv('../data/MachineLearningRating_v3.csv')

C:\Users\Tsebaot\AppData\Local\Temp\ipykernel_9060\3627243033.py:1: DtypeWarning: Columns (32,37) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('../data/MachineLearningRating_v3.csv')


## Risk Differences Across Provinces

### Analysis Objective
This test examines whether there are significant differences in **risk levels** (measured by `Total Claims`) across provinces. The goal is to understand how risk varies regionally, which can inform province-specific policies or risk management strategies.

### Hypotheses
- **Null Hypothesis (H₀):** No risk differences across provinces.
- **Alternative Hypothesis (H₁):** Risk differences exist across provinces.

In [9]:
# Test risk differences across provinces
test_risk_differences_across_provinces(df)


ANOVA Test for Risk Differences Across Provinces:
Null Hypothesis: No risk differences across provinces
F-Statistic: 5.849413762407606
P-Value: 1.6782057588675903e-07
Result: reject the null hypothesis



### Results
- **Test Type:** ANOVA (Analysis of Variance)
  - Compares the mean `Total Claims` across multiple provinces.
- **F-Statistic:** 5.849
  - Indicates that the variance in `Total Claims` between provinces is significantly greater than the variance within provinces.
- **p-Value:** 1.678205
  - This value is much smaller than the common significance level of 0.05, suggesting the observed differences are highly unlikely to be due to random chance.
- **Decision:** Reject the null hypothesis (H₀).

### Conclusion
There is strong evidence to conclude that **significant risk differences exist across provinces**.

### Implications
1. **Risk Management:** Provinces with higher average claims may require stricter risk mitigation measures, while lower-risk provinces could benefit from premium reductions.
2. **Pricing Strategy:** Develop province-specific premium structures to reflect the risk profile of each province.
3. **Further Analysis:** Investigate the factors contributing to risk differences, such as demographic, geographic, or economic factors.



## Risk Differences Between Zip Codes

### Analysis Objective
This test examines whether there are significant differences in **risk levels** (measured by `Total Claims`) between zip codes. The goal is to evaluate how risk varies geographically at a finer level, providing insights for localized strategies.

### Hypotheses
- **Null Hypothesis (H₀):** No risk differences between zip codes.
- **Alternative Hypothesis (H₁):** Risk differences exist between zip codes.

In [14]:
# Test risk differences between two zip codes 
test_risk_differences_between_zipcodes(df, 1000, 2000)



T-Test for Risk Differences Between Zip Codes 1000 and 2000:
Null Hypothesis: No risk differences between zip codes
T-Statistic: -0.6196332744567253
P-Value: 0.5355003097164719
Result: fail to reject the null hypothesis


### Results
- **Test Type:** T-Test (Analysis of Variance)
  - Compares the mean `Total Claims` across multiple zip codes.
- **T-Statistic:** -0.6196
  - Suggests a relatively low variance in `Total Claims` between zip codes compared to within zip codes.
- **p-Value:** 0.5355
  - This value is greater than the common significance level of 0.05, indicating that the observed differences are likely due to random chance.
- **Decision:** Fail to reject the null hypothesis (H₀).

### Conclusion
There is **insufficient evidence** to conclude that significant risk differences exist between zip codes. Any observed differences in `Total Claims` are likely due to random variation.

### Implications
1. **Uniform Risk Strategy:** Since risk levels appear consistent across zip codes, there may be no need for differentiated risk management or pricing strategies based on zip codes.
2. **Further Exploration:** Investigate other variables (e.g., demographics, policy types) that may reveal meaningful differences between zip codes.
3. **Combination Analysis:** Consider analyzing zip code risk differences alongside other factors, such as region or gender, for a more comprehensive understanding.

### Next Steps
- Visualize `Total Claims` by zip code to confirm the absence of meaningful trends.
- Conduct subgroup analyses (e.g., zip code and gender combinations) to uncover any hidden patterns.


## Margin Differences Between Zip Codes

### Analysis Objective
This test examines whether **profit margins** (calculated as `Premium - Total Claims`) differ significantly across zip codes. Profit margin reflects the profitability of policies in each zip code, making this analysis critical for identifying underperforming or highly profitable areas.

### Key Difference from Risk Differences Test
- **Test Risk Between Zip Codes:**
  - Focuses on analyzing **Total Claims** to determine if there are significant risk differences between zip codes.
  - Guides **risk management** strategies by identifying regions with higher or lower claims.
- **Test Margin Differences Between Zip Codes:**
  - Focuses on analyzing **Profit Margins** (`Premium - Total Claims`) to determine if there are significant profitability differences between zip codes.
  - Guides **pricing and profitability optimization** by highlighting regions that are overperforming or underperforming.

### Hypotheses
- **Null Hypothesis (H₀):** No significant margin differences between zip codes.
- **Alternative Hypothesis (H₁):** Significant margin differences exist between zip codes.

In [ ]:
# Test profit differences between two zip codes 
test_profit_differences_between_zipcodes(df, 1000, 2000)


T-Test for Profit Differences Between Zip Codes 1000 and 2000:
Null Hypothesis: No significant margin differences between zip codes
T-Statistic: 0.4776384393714926
P-Value: 0.6329083469646521
Result: fail to reject the null hypothesis


### Results
- **T-Statistic:** 0.4776
  - Indicates a smalle  variance in profit margins across zip codes compared to within-group variance.
- **p-Value:** 0.6329
  - This value is greater than the common significance level of 0.05, indicating that the observed differences are likely due to random chance.
- **Decision:** Fail to reject the null hypothesis (H₀).

### Conclusion
There is **insufficient evidence** to conclude that **significant profit margin differences exist between zip codes**.
Profits generated in these two zip codes are statistically similar.


### Implications
1. From a profit perspective, there is no need to differentiate strategies for these two zip codes.





## Interpretation: Risk Differences Between Genders

### Analysis Objective
This test examines whether there are significant differences in **risk levels** (measured by `Total Claims`) between genders. Understanding risk differences by gender can help insurers design gender-specific policies or adjust premiums based on claims data.

### Hypotheses
- **Null Hypothesis (H₀):** No significant risk differences between women and men.
- **Alternative Hypothesis (H₁):** Significant risk differences exist between women and men.

In [16]:
# Test risk differences between Women and Men
test_risk_differences_between_genders(df)


T-Test for Risk Differences Between Women and Men:
Null Hypothesis: No significant risk differences between women and men
T-Statistic: 0.24803623812388725
P-Value: 0.8041073961270343
Result: fail to reject the null hypothesis


### Results
- **Test Type:** T-Test (Independent Samples)
  - Compares the mean `Total Claims` between two independent groups: women and men.
- **T-Statistic:** 0.248
  - Indicates the magnitude of the difference between the means relative to the variability within groups.
- **p-Value:** 0.8041
 - This value is greater than the common significance level of 0.05, indicating that the observed differences are likely due to random chance.
- **Decision:**  Fail to reject the null hypothesis (H₀).

### Conclusion
Risk profiles are similar across genders.

### Implications
1. Gender-based differences do not appear to influence risk. Policies and strategies can remain gender-neutral for risk management purposes.